### PR TITLE
add Kernel Inception Distance to tf.contrib.gan.eval

### PR DESCRIPTION
**Pull request by djsutherland**
The KID is a score similar to the FID, but with an unbiased, asymptotically normal estimator. It was introduced by our paper [_Demystifying MMD GANs_](https://arxiv.org/abs/1801.01401); a very similar metric was also recommended by [_An empirical study on evaluation metrics of generative adversarial networks_](https://arxiv.org/abs/1806.07755).

For comparison to the FID, see section 4 (starting page 7) and appendices D/E (starting page 30) of our paper. As noted in the docstrings for the FID here, the FID estimator is biased and you can absolutely only compare estimates based on the same number of samples. But there's no guarantee that you still won't be very misled by the bias even then; in particular, see our Appendix D.2:

> This example thus gives a case where, for the dimension and sample sizes at which we actually apply the FID and for somewhat-realistic distributions, comparing two models based on their FID estimates will not only not reliably give the right ordering – with relatively close true values and high dimensions, this is not too surprising – but, more distressingly, will _reliably give the wrong answer_, with misleadingly small variance. This emphasizes that unbiased estimators, like the natural KID estimator, are important for model comparison.

Compared to our original code ([here](https://github.com/mbinkowski/MMD-GAN/blob/master/gan/compute_scores.py)), this version uses a slightly different estimator. Both are unbiased, but I think this block variant is more intuitive for general usage; its output is also "more normal" and gives a very simple estimate of the variance of the estimator (unlike the asymptotic one we used before, which is [not a very pretty expression](https://github.com/mbinkowski/MMD-GAN/blob/master/gan/compute_scores.py#L251)).

The functions that estimate the variance currently return an estimate as long as there are at least two blocks. This could be a little misleading; it might make sense to refuse to estimate the std if there are fewer than, say, 10 blocks, but I don't know if that added code complexity is worth it.

xref: https://github.com/google/compare_gan/pull/7

### CODE REVIEW

**Comment by goodfeli** [comment]
I'm not trained to be a TensorFlow reviewer, but I just wanted to comment that this is an important feature, contributed by the authors of the paper that introduced the method. My team is likely to use this metric in several of our research papers and we'd like it to be added to TensorFlow to encourage the general community to use it as well. I'm a co-author of the existing "Inception Score" metric that is already in TensorFlow, and I believe researchers should abandon Inception Score and switch to the new metric.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ispirmustafa** [comment]
@joel-shor could you please take a look?

**Comment by joel-shor** [comment]
sorry for the delay; im on vacation. ill review this next week

File: tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py [hunk 1]
```diff
@@ -734,3 +741,372 @@ def frechet_classifier_distance_from_activations(real_activations,
...
+
+  The estimator used takes time quadratic in max_block_size. Larger values of
+  max_block_size will decrease the variance of the estimator but increase the
+  computational cost. This differs slightly from the estimator used by the
+  original paper; it is the block estimator of https://arxiv.org/abs/1307.1954.
+
+  NOTE: the blocking code assumes that real_activations and
+￼ generated_activations are both in random order. If real_activations is sorted
+￼ in a meaningful order, the estimator will be biased.
+
```
Comment (joel-shor) [inline tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py]: nit: Is the indentation here correct?

File: tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py [hunk 2]
```diff
@@ -734,3 +741,372 @@ def frechet_classifier_distance_from_activations(real_activations,
...
+  estimates across models much more intuitive.
+
+  The estimator used takes time quadratic in max_block_size. Larger values of
+  max_block_size will decrease the variance of the estimator but increase the
+  computational cost. This differs slightly from the estimator used by the
+  original paper; it is the block estimator of https://arxiv.org/abs/1307.1954.
+
+  NOTE: the blocking code assumes that real_activations and
+￼ generated_activations are both in random order. If real_activations is sorted
+￼ in a meaningful order, the estimator will be biased.
```
Comment (joel-shor) [inline tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py]: Same

**Review by joel-shor — DISMISSED** [review]

File: tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py (hunk 2 above)
Comment (joel-shor) [reply tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py]: Great addition. Thanks!

Comment (dougalsutherland) [reply tensorflow/contrib/gan/python/eval/python/classifier_metrics_impl.py]: Fixed these, thanks.

**Comment by dougalsutherland** [comment]
Hey @joel-shor @ispirmustafa, just checking in about this. :)

**Review by joel-shor — DISMISSED** [review]

**Comment by dougalsutherland** [comment]
Test failures were mostly from a strange non-ascii whitespace character; no idea how that snuck in, something weird with github's web editor maybe? The only other one I noticed was an XLA error that clearly had nothing to do with this PR.

**Review by joel-shor — DISMISSED** [review]

**Comment by joel-shor** [comment]
Please wait for the tests to come back green, then we can submit.

Thanks for your patience, and thanks for the excellent contribution!!

**Comment by dougalsutherland** [comment]
Sorry, the MacOS Contrib failure was a legit mistake in the test code: `math.ceil` is apparently a float in py2 even though it's an int in py3. One more time. :/

**Comment by dougalsutherland** [comment]
Sorry @joel-shor -- can you run the tests again? Should be fixed now.

**Comment by dougalsutherland** [comment]
@joel-shor @ispirmustafa, can you add the label to run the tests again? They should pass now.

**Review by joel-shor — DISMISSED** [review]

**Comment by dougalsutherland** [comment]
`keras_test` timeouts and the XLA error have nothing to do with this code, and the Ubuntu Python3 tests that infrastructure-errored passed on the previous commit, FYI. So it seems like tests essentially passed. :)

**Comment by dougalsutherland** [comment]
Just for the record, now: the XLA error is clearly unrelated, and so is the Windows Bazel GPU error, which is:

```
error C3083: '`global namespace'': the symbol to the left of a '::' must be a type
external/com_google_absl\absl/container/inlined_vector.h(85): note: while compiling class template member function 'absl::InlinedVector<tensorflow::TensorReference,4,std::allocator<T>>::InlinedVector(void) noexcept'
        with
        [
            T=tensorflow::TensorReference
        ]
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dougalsutherland** [comment]
Rebased to current `master` in case the other errors have been fixed in the meantime. @joel-shor, can you hit run again?

**Review by joel-shor — APPROVED** [review]

**Comment by dougalsutherland** [comment]
Ping; again, the test failures clearly have nothing to do with this code.

**Comment by dougalsutherland** [comment]
:tada: Thanks for handling this @joel-shor!
